go/ql/lib/semmle/go/security/UnicodeBypassValidationQuery.qll

```diff
@@ -0,0 +1,149 @@
+  exists(DataFlow::CallNode cn, DataFlow::EqualityTestNode etn |
+    g = etn and
+    DataFlow::localFlow(cn.getResult(), etn.getAnOperand()) and
+    cn.getTarget()
+        .hasQualifiedName("strings",
+          [
+            "Index", "IndexAny", "IndexByte", "IndexFunc", "IndexRune", "LastIndex", "LastIndexAny",
+            "LastIndexByte", "LastIndexFunc", "Count",
+          ]) and
+    cn.getArgument(0).asExpr() = e and
+    etn.getAnOperand().getIntValue() = -1 and
+    outcome = etn.getPolarity()
+  )
+}
+
+private predicate boolCheck(DataFlow::Node cn, Expr e, boolean outcome) {
+  cn.(DataFlow::CallNode)
+      .getTarget()
+      .hasQualifiedName("strings",
+        ["Contains", "ContainsAny", "ContainsRune", "HasPrefix", "HasSuffix", "EqualFold"]) and
+  cn.(DataFlow::CallNode).getArgument(0).asExpr() = e and
+  outcome = false
+}
+
+/**
+ * A call to a function called `Index`, `ContainsAny`, or similar, which may be
+ * considered a barrier guard for eliminating untrusted Unicode characters.
+ */
+class UntrustedUnicodeCharChecks extends DataFlow::Node {
+  UntrustedUnicodeCharChecks() {
+    this = DataFlow::BarrierGuard<intCheck/3>::getABarrierNode()
+    or
+    this = DataFlow::BarrierGuard<boolCheck/3>::getABarrierNode()
+  }
+}
+
+/**
+ * A taint-tracking configuration for detecting "Unicode transformation mishandling" vulnerabilities.
+ *
+ * This configuration uses two flow states, `PreValidation` and `PostValidation`,
+ * to track the requirement that a logical validation has been performed before the Unicode Transformation.
+ */
+class Configuration extends TaintTracking::Configuration {
+  Configuration() { this = "UnicodeBypassValidation" }
+
+  override predicate isSource(DataFlow::Node source, DataFlow::FlowState state) {
+    (source instanceof UntrustedFlowSource or source instanceof Source) and
+    state instanceof PreValidation
+  }
+
+  override predicate isSanitizer(DataFlow::Node sanitizer, DataFlow::FlowState state) {
+    sanitizer instanceof Sanitizer and
+    state instanceof PostValidation
+  }
+
+  override predicate isAdditionalTaintStep(
+    DataFlow::Node nodeFrom, DataFlow::FlowState stateFrom, DataFlow::Node nodeTo,
+    DataFlow::FlowState stateTo
+  ) {
+    (
+      exists(DataFlow::CallNode cn |
+        cn.getACalleeIncludingExternals().asFunction() instanceof EscapeFunction and
+        nodeFrom = cn.getAnArgument() and
+        nodeTo = cn.getResult()
+      )
+      or
+      exists(DataFlow::CallNode cn |
+        (
+          cn.getACalleeIncludingExternals().asFunction() instanceof RegexpMatchFunction or
+          cn.getACalleeIncludingExternals().asFunction() instanceof RegexpReplaceFunction
+        ) and
+        nodeFrom = cn.getAnArgument() and
+        nodeTo = cn.getResult()
+      )
+      or
+      exists(DataFlow::CallNode cn |
+        cn.getTarget().hasQualifiedName("strings", "Compare") and
+        nodeFrom = cn.getAnArgument() and
+        nodeTo = nodeFrom
```

Comment:
This looks like a case for a barrier, not an additional step

Reply:
indeed, the Regex match function calls are also a case of a barrier, commit https://github.com/github/codeql/pull/12994/commits/2a22decce96b4b884e1964823b8a6b55dabeeca0.